### DISCUSSION TITLE
with immer midderware, Array type wrong

### DISCUSSION THREAD

**Discussion by alpacachen** [pmndrs/zustand#2152] [Bug report] [answered]
an immer store, can't use Arrray.sort, but TS check escaped.

Error in /~/src/App.tsx (23:62)
Cannot assign to read only property '0' of object '[object Array]'

demo : https://stackblitz.com/edit/stackblitz-starters-ebbhts?file=src/App.tsx

hope TS checker can emit error correctly

**Comment by dai-shi** [comment]
Is it `push` instead of `sort` in the demo?

**Reply by dai-shi** [reply]
Oh, I found it commented out.

So, is it a typescript issue? Or, a runtime issue?

**Reply by alpacachen** [reply]
runtime issue，in demo ，click inc button twice，can getthe error

**Reply by alpacachen** [reply]
I think its better to expose error during ts
Check

**Reply by dai-shi** [reply]
https://codesandbox.io/s/zealous-hodgkin-k3f6nz?file=/src/App.tsx
Disabling auto freeze seems to work around.

**Reply by dai-shi** [reply]
Or, copy the array.
https://codesandbox.io/s/dreamy-brahmagupta-448h98

**Comment by dai-shi** [comment]
Not sure if anything can be done on zustand end.
Please feel free to investigate it if anyone is interested.

Here's the code without types: https://unpkg.com/browse/zustand@4.4.4/esm/middleware/immer.mjs
<img width="926" alt="image" src="https://github.com/pmndrs/zustand/assets/490574/a068fdef-1b2f-4b5b-b799-d5740d1793ed">

**Comment by alpacachen** [comment]
> Or, copy the array. https://codesandbox.io/s/dreamy-brahmagupta-448h98

Im not sure its a good way,because i can not Ask my colleagues to do the same

**Comment by alpacachen** [comment]
> https://codesandbox.io/s/zealous-hodgkin-k3f6nz?file=/src/App.tsx Disabling auto freeze seems to work around.

this is the good way

**Comment by alpacachen** [comment, marked as answer]
https://immerjs.github.io/immer/typescript#best-practices

Immer Best practices
